content: draft: define source-track objective in terms of revisions and provenance. (#1083)
slsa-framework/slsa: docs/spec/v1.1/source-requirements.md

## Objective

The SLSA Source Track mitigates [Threat A ("Submit unauthorized change")](/spec/v1.0/threats#a-submit-unauthorized-change), scoped to a code repository and the organization that owns that repository. Concretely: an attacker must compromise the accounts of two organization members to publish code in a Source Level 3-conformant repository, and the evidence of those unauthorized changes cannot be destroyed without further attacks.

## Changes from v0.1

-   **Scope** The Source track is now scoped to Revisions rather than builds.
Why?: To facilitate verification without anchoring it to a build.

-   **Model** Added a model, definitions, and the concept of verification.
Why?: SLSA does not yet have a model for version control systems, and we need such a model to be able to discuss them.

> ## Outstanding TODOs

Author: happy to add these back, it feels like this content should be (and is?) tracked in issues?

Reviewer: I think the idea is that while the source track is still in draft form it's helpful to keep the TODOs here as an indication it's not quite ready yet.

Author: how do we feel about 

```
status: draft
open issues: https://github.com/slsa-framework/slsa/issues?q=is%3Aissue%20state%3Aopen%20label%3Asource-track
```

or something else that keeps it dynamic?

Reviewer: SGTM

Reviewer 2: historically we've biased towards leaving todos in the doc so we don't have to context switch away from the editor to work through them, I think it would be useful to leave them/add them back.

Author: this PR will be rebased onto the results of : https://github.com/slsa-framework/slsa/pull/1088/files. 

I'm still not sure I see the utility of keeping these stale topics under VCS, I'll refresh the list when this is put back up for review.

> The SLSA SOURCE track describes increasing levels of trustworthiness and completeness in a repository revision's provenance. 

Reviewer: open question: should we call this 'provenance' (which is overloaded with the specific format used by SLSA to track _build_ information) or should we call it something else?

If we do continue to use the word 'provenance' here we'll need to update other documentation which only uses 'provenance' when talking about mapping build -> source (and not when talking about source itself).  E.g. https://slsa.dev/spec/v1.0/provenance

We may be able to work around this by eliding the word entirely?

> The SLSA SOURCE track describes increasing levels of trustworthiness and completeness in repository revisions. and their history.
>
> This history describes which Source Control Provider produced the revision, what process was used, and who were the contributors. 
> The lowest level only requires the provenance to exist, while higher levels provide increasing protection against tampering of the version control system, the provenance, or the revision.
> ...

Reviewer 3: I suggest we use a different word to avoid confusion.

Author: @TomHennen @adityasaky I think the objective section of this doc should settle on the term for "the process that created a revision."

I think mechanically this thing would be a "revision creation process attestation" so generally the word provenance seems to work pretty well and carries the right semantics. 
If we choose a new term, it would need to make sense on its own too: You'd fetch the "provenance attestation" for a file artifact and the "<new term> attestation" for the repo artifact, etc. 

Potentially we could just use: "predicateType": "https://slsa.dev/source-provenance/v1" or something else to differentiate the schema. Conceptually, these schemas would hold very similar data, so it might be that most of the existing docs would hold up unless they were discussing specific data elements?

Reviewer: You've convinced me.  How does being deliberate about referring to it as "source provenance" in this track (and then updating the build docs to talk about "build provenance" in another PR) sound?

We can then see how it's received later.

Reviewer 3: I think settling on the term for the process is fine. I also agree that we could define a separate predicate type; I'm not too worried about confusion when this is implemented in attestors / verifiers, but rather in regular conversations where context may be muddier. But let's punt that!

Reviewer 2: +1 to using "source provenance" and "build provenance" as a starting point.

> The SLSA SOURCE track describes increasing levels of trustworthiness and completeness in a repository revision's provenance. 

Reviewer 2: ```suggestion
The SLSA source track describes increasing levels of trustworthiness and completeness in a repository revision's _provenance_. 
```

This matches the style in https://slsa.dev/spec/v1.0/levels

> Provenance describes which Source Control Provider produced the revision, what process was used, and who were the contributors. 

Reviewer 2: ```suggestion
Provenance describes which Source Control Platform (SCP) produced the revision, what process was used, and who were the contributors. 
```

> The lowest level only requires the provenance to exist, while higher levels provide increasing protection against tampering of the version control system, the provenance, or the revision.

Author: this PR doesn't actually address the leveling definitions further on in the document.

Reviewer 3: is the idea to add more to this PR? I think it might be fine to handle them in separate PRs to keep diffs small and easier to reason about and discuss.

Reviewer: I think the idea is to handle the level definitions in another PR.

> The lowest level only requires the provenance to exist, while higher levels provide increasing protection against tampering of the version control system, the provenance, or the revision.

Reviewer 2: ```suggestion
The lowest level only requires the provenance to exist, while higher levels provide increasing protection against tampering with any parts of the Source Control Platform, the provenance, or the revision.
```

Author: Probably we can cut this line altogether, but as written the "lowest level" wouldn't involve SCP yet.

> Consumers have some way of knowing what the expected provenance should look like for a given package and then compare each source revision's actual provenance to those expectations. 

Reviewer: > The primary purpose of the source track is to enable verification that a revision followed the expected process. 
Consumers have some way of knowing what the expected provenance should look like for a given package and then compare each source revision's actual provenance to those expectations. 

I wonder if we can cut this section?

IIUC it's modified from [the 'Build track' description](https://slsa.dev/spec/v1.0/levels#build-track:~:text=Consumers%20have%20some%20way%20of%20knowing%20what%20the%20expected%20provenance%20should%20look%20like%20for%20a%20given%20package%20and%20then%20compare%20each%20package%20artifact%E2%80%99s%20actual%20provenance%20to%20those%20expectations.%20Doing%20so%20prevents%20several%20classes%20of%20supply%20chain%20threats.), but the interesting thing about the build track is that there's much greater variability in the way a build process is done, so the 'expected' process needs to be defined and communicated somehow and that process may vary from project to project.

However, I think 'source' is different?  The 'expected process' should have much less variability and it's easier to introspect?  Instead of expected process I think we're looking for something like "the source was written or approved of by the project's maintainers" (or something like that)

Author: I was initially thinking that too but have kind of come around. 
While in practice SLSA may make a few specific recommendations (such as around code review process), virtually all consumers will be interested in whether their own policy CI jobs were run properly on the revision (github has _many_ of these). 

To phrase it differently, is the expected verification flow "ensure that all the configured rules passed?" or "ensure that the specific rules I care about passed?" 

From experience, it tends to be a mix: most of the time customers care that "the configured rules were followed" and sometimes they need to know that "definitely this revision passed a specific policy-related-ci."

I feel like this aligns pretty well to an "expectation matching" approach, similar to the build track.

Reviewer: Oh hmm.  I think we have different ideas in mind here.

I was thinking the source track would identify rules that _must_ be enforced for each level and be silent on any others.  That we'd leave any other rules that might need to be checked to something else.  (But I definitely understand that use case).  The way I'd expect this to work is that downstream consumers of this source probably won't want to care about the specific rules but the repo owner may, and so I'd be inclined to leave this to some sort of policy evaluation that can happen on the repo (the same thing that makes the determination for downstream users).

Note that I _definitely_ see the value in this sort of thing, but given that I think it can be handled separately, and that it would make the source track more complicated (at least I expect it would) how would you feel about deferring some of that pluggability until later (or some orthogonal effort)?

Do you have any particular rules in mind?

Reviewer 3: Given that this is specifically about the _provenance_ about the source changes, I lean towards just capturing what process was followed and leaving it to the verifier to set expectations. So, effectively "ensure that the specific rules I care about passed" in @zachariahcox's comment above.

Source track is indeed a bit different from the build track but IMO there's value in keeping the separation between recording what happened (the source provenance) and what the expectations are when verifying the source provenance. It allows for greater agility wrt which parties we trust for what data. We also don't have to reason about how different expectations measure up against each other. For example, the maintainers may have expectations set A and I may have expectations set B, where B is in fact a subset of A. Instead of reasoning about that for every individual security control making up A and B, we can independently verify A and B are met by inspecting the provenance rather than having this extra bit of indirection.

> Note that I definitely see the value in this sort of thing, but given that I think it can be handled separately, and that it would make the source track more complicated (at least I expect it would) how would you feel about deferring some of that pluggability until later (or some orthogonal effort)?

I know this question is for @zachariahcox but I'm not sure I fully understand this. Could you clarify, perhaps with an example, what you propose we defer? 😄

Author: Great discussion. I definitely meant @TomHennen 's point 3: how many tools will a customer need to use to verify something? Or baking in an assumption that the attestation format will at least be somewhat extensible.

For teams that need more, I expect the SCP will want to add the additional claims in there somewhere (maybe akin to how the [gh oidc token contains a bunch of useful, gh-specific info](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token)). 

Empirically, to satisfy at least some customers, the revision-creation-process-attestation would need to contain no less than all the info captured by [ruleset insights](https://youtu.be/BYajh38HhCs?t=1025), plus a qualitative assessment as to whether those configured rulesets _also_ combine to fulfill any SLSA levels. 

> how would you feel about deferring some of that pluggability until later (or some orthogonal effort)?

💯 👍 
Does that effectively mean that for the first pass, we're _only_ targeting summaries?
Even with that goal, the wording like "Consumers have some way of knowing what the expected provenance should look like ... and compare it to actual" is pretty accommodating.

Reviewer: Oh I see, thanks!  So I think I have a good idea on how to handle nicely.  Perhaps I should put together a proposal for the consumer side attestation that would do what you're looking for?

How do you feel about "Consumers have some way of knowing what security properties the revision meets and how they can trust that determination".

But now that I say that I also see that in the original wording "Consumers have some way of knowing what the expected provenance should look like for a given package and then compare each source revision's actual provenance to those expectations. " I might have gotten confused with the two different mentions of "provenance" :).

I think what that is trying to say is "The primary purpose of the source track is to enable verification that _the process to create a revision meets the expected security properties_.  Consumers have some way of mapping the sources listed in the _build provenance_ to the various _source provenance_ attestations to determine if all sources used during the build meet their requirements."

Author: @TomHennen this comment got disconnected post rebase. 👀

Reviewer: Found it and made a suggestion...

## Definitions

| Term | Description|
| --- | — |

> | Subject | The subject of the attestation is the immutable identifier of a repository revision. EG: https://github.com/slsa-framework/slsa/commit/306642f21dbdaca2eaafe2df8e98432b4d4f2f02

Reviewer: This seems like an interesting topic of conversation.  When I read 'subject' here I'm thinking of the subject field in [the in-toto statement layer](https://github.com/in-toto/attestation/blob/main/spec/v1/statement.md) and I'd imagine that to look a lot like [`resolvedDependencies` does](https://slsa.dev/spec/v1.0/provenance#:~:text=%22resolvedDependencies%22%3A%20%5B%7B%0A%20%20%20%20%22uri%22%3A%20%22git%2Bhttps%3A//github.com/octocat/hello%2Dworld%40refs/heads/main%22%2C%0A%20%20%20%20%22digest%22%3A%20%7B%22gitCommit%22%3A%20%227fd1a60b01f91b314f59955a4e4d4e80d8edf11d%22%7D%0A%7D%5D) so that the build provenance and source attestation can easily be matched.

So I suppose my question is: can we update the example or just remove it until we're more firm on what the attestation format actually is?

Author: Yeah, we can remove the specific example for now. 
However, if we're basing this on revisions, that's just a url to a revision.

Reviewer: Ok, yeah let's remove for now and then discuss.  My biggest qualm about changing it from what's in resolvedDependencies is that to put it to use we'd have to change the build provenance spec which seems hard.

Reviewer 3: I think this could be a resource descriptor, like resolvedDependencies, indeed. Note that the attestation framework supports `gitCommit`, `gitTree` etc as types of hashes, so this should be quite straightforward to represent.

Author: > to put it to use we'd have to change the build provenance spec which seems hard

I'd guess that you can resolve a dependency that comes with its own provenance attestation though, right? 
Would this case be different?

> I think this could be a resource descriptor, like resolvedDependencies

Nice! Wouldn't this source provenance attestation still have a revision as its subject though?
I'm probably missing some nuance here.

Reviewer: Yes I think it would.

It would look something like

```
...
    "subject": [
        {
            "uri": "git+https://github.com/octocat/hello-world@refs/heads/main",
            "digest": {"gitCommit": "7fd1a60b01f91b314f59955a4e4d4e80d8edf11d"}
        }
    ]
...
```

Author: >  "uri": "git+https://github.com/octocat/hello-world@refs/heads/main",

this reminds me that we need to define "ref" or "branch" in the spec.

Reviewer: And that reminds me that there are cases where the destination branch / release is often an important part of the context for a 'review'.  So it might not be enough to say commit 'abc123' was reviewed, it might be necessary to know "reviewed in what context".

Author: yeah 💯 , that should be part of the "situation" in the change definition I think.

> | Source | An identifiable set of text and binary files and associated metadata.

Reviewer: Any particular reason to drop "usually used as input for the build system (see SLSA Build Track)."

Author: In the other issue there was a comment thread about there being tons of uses for version control system revisions. 

We could continue to point out that repos could be consumed by other processes.

Reviewer: Yeah that SGTM

| Organization | A collection of people who collectively create the Source. Examples of organizations include open-source projects, a company, or a team within a company. The organization defines the goals and methods of the repository.

> | Repository | A uniquely identifiable instance of a VCS hosted on an SCP. The repository controls access to the Source in the version control system. The objective of a repository is to reflect the intent of the organization that controls it.

Reviewer 2: We should move VCS and SCP definitions up so that we have defined terms before we use them.

> | Change | A set of modifications to one or more source files and associated metadata. A change is proposed by a “publisher,” and applied to specific revision to create a new revision. 

Reviewer 2: Do we need to define publisher? 🤔

Reviewer: I think this is obsolete with the new wording?

Change metadata MUST include any information required to situate the change in relation to other changes (e.g. parent revision).
| Version Control System (VCS)| Software for tracking and managing changes to source. Git and Subversion are examples of version control systems.
| Revision | A specific identifier provided by the version control system that identifies a given state of the source. As an example, you can identify a git revision by its tree hash.
| Change History | A record of the history of changes that went into the revision.
| Source Control Platform (SCP) | A service or suite of services for hosting version-controlled software. GitHub and GitLab are examples of source control platforms, as are combinations of tools like Gerrit code reviews with GitHub source control.

> | Source Provenance Attestation | A signed document describing the security claims associated with a revision.

Reviewer: This is probably related to my question about naming above.  We might consider leaving this definition out until we actually have the attestation defined?

I also wouldn't be surprised if we wound up with multiple attestation types.  One for consumers to understand the claim of a given source level in a revision and one that is used as evidence to compute the source level for all revisions.

Author: I think this is a specific issue digging into this topic https://github.com/slsa-framework/slsa/issues/758

I would support cutting it for now if it's at all contentious.

Reviewer: Yeah my preference would be to cut and then add back in when we have the actual attestation proposals.
